README.md
## Analysis of the DPOE 2014 Survey Results

### Introduction

### Contents

* dpoe14_survey.Rproj - default setup for R project

#### data - datasets to be analyzed
* data_clean.csv - a cleaned version of the survey responses. See change_log.txt (not currently included)
* data_original.csv - the orginal survey responses as exported from SurveyMonkey (not currently included)
* data_popbystate - state by state population data from the US census (not currently included)
* change-log.txt - a description of changes made while cleaning the data
* codebook.txt - description of the survey variables

#### analysis - R scripts for analyzing the data
* 0_functions.r - shared R functions for analysis
* 1_setup.r - load and classify of data
* 2_time.r - subset and graph time data
* 3_org.r - subset and graph organizational data
* 4_training - subset and graph training preferences
* scratch.r - experiment with data views

#### reports - markdown files for discussing analysis
* analysis-tufte.Rnw - source for report formatted according to the Tufte class


data/change_log.txt
20140111
-Dropped columns: RespondentID, CollectorID, IP Address, Email Address, First Name, LastName, Custom Data
-Converted ParentOrg*, Type*, StaffSize, Resp*, ContForm*, ContType*, Importance, AidAmount, and Notification to numerical categories
-Cleaned all Other rankings that had no comment. If no comment ranking was ranked greater than last, all other rankings were updated to take into account this deletion. For examples, a ranking vector of [4, 3, 2, 1, n/a] is updated to [3, 2, 1, n/a, n/a].

analysis/2_time.R
# subset of the data related to survey metadata
# extra columns calculated for finer grain time analysis
df.meta <- select(df, StartDateTZ, EndDateTZ, State, Timezone) %>%
        mutate(Length = as.numeric(EndDateTZ - StartDateTZ),
               # add day for week function since year started on a Tuesday
               StartWeek = factor(week(StartDateTZ + days(1))),
               StartWeekday = wday(StartDateTZ, label = T),
               StartHour = factor(hour(StartDateTZ)))

# outline = F to ignore outliers
par(bty = 'n')
a <-boxplot(df.meta$Length, outline = F, horizontal = T, col = '#B34F89', xlab = 'Minutes')

# count the survey responses by week
g.Weeks <- ggplot(df.meta, aes(x=StartWeek)) +
        geom_histogram(binwidth = 8, fill = '#B34F89') +
        theme(axis.title.x=element_blank(), 
              axis.title.y=element_blank())

# plot the total number of surveys taken
g.Date <- ggplot(df.meta, aes(x=StartDateTZ)) +
        stat_ecdf(col = '#B34F89', size = 1.5) +
        labs(x = "Date",
             y = "Total Percentage of Surveys")

# count the survey responses by week
g.DOW <- ggplot(df.meta, aes(x = StartWeekday)) +
        geom_histogram(fill = '#B34F89') +
        theme(axis.title.x=element_blank(), 
              axis.title.y=element_blank())

# count the survey responses by hour
g.hour <- ggplot(df.meta, aes(x=StartHour)) +
        geom_histogram(fill = '#B34F89') +
        labs(x = "Hour", 
             y = "Number of Surveys")

# run same plot with facets from timezones
g.hourTZ <- ggplot(df.meta, aes(x=StartHour)) +
        geom_histogram(fill = '#B34F89') +
        facet_grid(Timezone ~.) +
        labs(x = "Hour (EDT)", 
             y = "Number of Surveys")

analysis/1_setup.R
setwd('~/dev/DPOE14_survey/')

# load packages for data analysis
library(scales)
library(ggplot2)
library(lattice)
library(lubridate)
library(plyr)
library(dplyr)
library(reshape2)

# increase number length before scientific notation is used
options(scipen = 6)

# load functions created during the course of this analysis
source('~/dev/DPOE14_survey/analysis/0_functions.r')

# data frame named df because I'm lazy 
df <- read.csv('~/dev/DPOE14_survey/data/data_clean.csv', stringsAsFactors = T, na.string = c('NA', '', 'N/A'))

# convert dates to POSIX objects
df$StartDate <- mdy_hms(df$StartDate, truncated=1, tz='Europe/London')
df$EndDate <- mdy_hms(df$EndDate, truncated=1, tz='Europe/London')
df$Timezone <- df$State

# assign timezones based on state
levels(df$Timezone) <- list('America/New_York' = c('ME', 'NH', 'VT', 'MA', 'RI', 'CT', 'NY',
                                                   'PA', 'NJ', 'DE', 'MD', 'VA', 'DC', 'NC',
                                                   'SC', 'GA', 'FL', 'WV', 'OH', 'MI', 'IN',
                                                   'KY', 'PR'),
                            'America/Chicago' = c('WI', 'IL', 'TN', 'MS', 'AL', 'LA', 'AR',
                                                  'MO', 'IA', 'MN', 'ND', 'SD', 'NE', 'KS', 
                                                  'OK', 'TX'),
                            'America/Denver' = c('MT', 'WY', 'CO', 'NM', 'AZ', 'UT', 'ID'),
                            'America/Los_Angeles' = c('OR', 'WA', 'CA', 'NV'),
                            'America/Juneau' = c('AK'),
                            'Pacific/Honolulu' = c('HI'))

# normalize times according to timezones
a <- lapply(seq(length(df$StartDate)), function(x) with_tz(df$StartDate[x], tz = as.character(df$Timezone[x]))[1])
df$StartDateTZ <- Reduce('c', a)

b <- lapply(seq(length(df$EndDate)), function(x) with_tz(df$EndDate[x], tz = as.character(df$Timezone[x]))[1]) 
df$EndDateTZ <- Reduce('c', b)

# clean up and order factor names
df$AidAmount <- revalue(df$AidAmount, c('$0 - $250' = '$0-250', '$251 - $500' = '$251-500', '$501 - $750' = '$501-750',
                                        '$751 - $1,000' = '$751-1000', '$1,001 - $3,000' = '$1001-3000', 
                                        '$3,001 - $5,000' = '$3001-5000', 'More than $5,000' = '$5000+', 'I don\'t know' = 'Don\'t know'))
df$AidAmount <- factor(df$AidAmount, c('$0-250', '$251-500', '$501-750', '$751-1000', '$1001-3000', '$3001-5000', '$5000+', 'Don\'t know'))

df$StaffSize <- revalue(df$StaffSize, c('1 - 12' = '1-12', '13 - 24' = '13-24', '25 - 50' = '25-50',
                                        '51 - 200' = '51-200', '201 - 500' = '201-500', '501+'))
df$StaffSize <- factor(df$StaffSize, c('1-12', '13-24', '25-50', '51-200', '201-500', '501+'))

analysis/scratch.R


## Participants

```{r}
df.types <- select(df, ResponseID, starts_with('T.'), -T.Other) %>%
melt(id = 'ResponseID', variable.name = 'OrgTypeCat', value.name = 'OrgType', na.rm=T)
df.parentorgs <- select(df, ResponseID, starts_with('PO.'), -PO.Other) %>%
melt(id = 'ResponseID', variable.name = 'ParentOrgCat', value.name = 'ParentOrg', na.rm=T)
df.staff <- select(df, ResponseID, starts_with('R.'), -R.Other) %>%
melt(id = 'ResponseID', variable.name = 'Staffing', value.name = 'Staffing', na.rm=T)

df.org <- select(df, ResponseID, State, Timezone, StaffSize, AidAmount)

#        merge(df.types[, c(1, 3)]) %>%
#        merge(df.staff[, c(1, 3)]) %>%
#        merge(df.parentorgs[, c(1, 3)])
```

```{r}
df.cformat <- select(df, ResponseID, starts_with('CF.'), -CF.Other) %>%
melt(id = 'ResponseID', variable.name = 'ContentFormatCat', value.name = 'ContentFormat', na.rm=T)
df.ctypes <- select(df, ResponseID, starts_with('CT.'), -CT.Other) %>%
melt(id = 'ResponseID', variable.name = 'ContentTypeCat', value.name = 'ContentType', na.rm=T)

df.content <- select(df, ResponseID, Importance)

```

## Training Preferences
```{r}
df.rtopic <- select(df, ResponseID, starts_with('RT.'), -ends_with('Other'), -ends_with('OtherComment')) %>%
melt(id = 'ResponseID', variable.name = 'Topic', value.name = 'TopicRanking') %>%
mutate(TopicRanking = factor(TopicRanking))
df.rformat <- select(df, ResponseID, starts_with('RTF.'), -ends_with('Other'), -ends_with('OtherComment')) %>%
melt(id = 'ResponseID', variable.name = 'Format', value.name = 'FormatRanking') %>%
mutate(FormatRanking = factor(FormatRanking))
df.rinperson <- select(df, ResponseID, starts_with('RTIP.'), -RTIP.PtherComment, -RTIP.OtherComment) %>%
melt(id = 'ResponseID', variable.name = 'InPersonFormat', value.name = 'InPersonFormatRanking') %>%
mutate(InPersonFormatRanking = factor(InPersonFormatRanking))
df.rinperson2 <- meltRanks(df, 'RTIP.', 'InPersonFormat', 'InPersonFormatRanking')
df.rlength <- select(df, ResponseID, starts_with('RTL.'), -RTL.Other, -RTL.OtherComment) %>%
melt(id = 'ResponseID', variable.name = 'Length', value.name = 'LengthRanking') %>%
mutate(LengthRanking = factor(LengthRanking))
df.rlength2<- meltRanks(df, 'RTL.', 'Length', 'LengthRanking')
df.routcome <- select(df, ResponseID, starts_with('RN.'), -RN.Other, -RN.OtherComment) %>%
melt(id = 'ResponseID', variable.name = 'Outcome', value.name = 'OutcomeRanking') %>%
mutate(OutcomeRanking = factor(OutcomeRanking))

df.routcome.top <- df.routcome[df.routcome$OutcomeRanking == 1, c(1,2)]

df.rtopic.routcome <- merge(df.rtopic, df.routcome.top, id = 'ResponseID')

ggplot(df.rtopic.routcome, aes(x = TopicRanking)) + 
geom_histogram(fill = '#B34F89') + 
facet_grid(Topic~Outcome)

ggplot(df.rtopic.routcome) + 
        geom_bar(aes(TopicRanking, (..count..)/sum(..count..)), fill = '#B34F89') + 
        scale_y_continuous(labels = percent) + 
        facet_grid(. ~ Topic + Outcome)
```

```{r}
source('functions.R')
df2<-topSlice(df.routcome, df.rformat, 'ResponseID')
topSliceHist(df2, 'FormatRanking', 'Format', 'Outcome')

df.routcome.top <- df.routcome[df.routcome$OutcomeRanking == 1, c(1,2)]

df.rformat.routcome <- merge(df.rformat, df.routcome.top, id = 'ResponseID')

ggplot(df.rformat.routcome, aes(x = FormatRanking)) + 
        geom_histogram(fill = '#B34F89') + 
        facet_grid(Format~Outcome)

ggplot(df.rtopic.routcome) + 
        geom_bar(aes(TopicRanking, (..count..)/sum(..count..)), fill = '#B34F89') + 
        scale_y_continuous(labels = percent) + 
        facet_grid(. ~ Topic + Outcome)
```

```{r}
ggplot(df.rlength, aes(x = LengthRanking)) + 
        geom_histogram() +
        facet_grid(Length ~ .)

df.routcome.top <- df.routcome[df.routcome$OutcomeRanking == 1, c(1,2)]

df.rlength.routcome <- merge(df.rlength, df.routcome.top, id = 'ResponseID')

ggplot(df.rlength.routcome, aes(x = LengthRanking)) + 
        geom_histogram(fill = '#B34F89') + 
        facet_grid(Length~Outcome)

df.rtopic.top <- df.rtopic[df.rtopic$TopicRanking == 1, c(1,2)]

df.rlength.rtopic <- merge(df.rlength, df.rtopic.top, id = 'ResponseID')

ggplot(df.rlength.rtopic, aes(x = LengthRanking)) + 
        geom_histogram(fill = '#B34F89') + 
        facet_grid(Length~Topic)

df.rformat.top <- df.rformat[df.rformat$FormatRanking == 1, c(1,2)]

df.rlength.rformat <- merge(df.rlength, df.rformat.top, id = 'ResponseID')

ggplot(df.rlength.rformat, aes(x = LengthRanking)) + 
        geom_histogram(fill = '#B34F89') + 
        facet_grid(Length~Format)
```

```{r}
ggplot(df.rformat, aes(x = FormatRanking)) + 
        geom_histogram() +
        facet_grid(Format ~ .)

select(df, starts_with('RTF.'), -RTF.Other, -RTF.OtherComment) %>%
        colMeans(na.rm = T)
```



b<-merge(df.parentorgs, df.ctypes)

ddply(b, .(ContentType), summarize, per = count(ContentType)/count(ResponseID))


analysis/0_functions.R
library(plyr)
library(dplyr)
library(reshape2)
library(ggplot2)

topSlice <- function (df, df.merge, idVar) {
        df.top <- df[df[, 3] == 1, c(1,2)]
        df.slice <- merge(df.top, df.merge, id = idVar)
}

topSliceHist <- function (df, df.merge, ranking, xFacet, yFacet) {
        df.top <- df[df[, 3] == 1, c(1,2)]
        df.slice <- merge(df.top, df.merge, id = idVar)
        
        ggplot(df.slice, aes_string(x = ranking)) + 
                geom_histogram(fill = '#B34F89') +
                facet_grid(as.formula(paste(xFacet, '~', yFacet)))
}

facetHist <- function (df, plotx, yFacet, xFacet) {
        g.plot <- ggplot(df, aes_string(x = plotx)) +
                geom_histogram(fill = '#B34F89') +
                theme(axis.title.x=element_blank(), 
                      axis.title.y=element_blank()) 
        
        ifelse(!missing(yFacet) & !missing(xFacet),
               g.plot <- g.plot + facet_grid(as.formula(paste(yFacet, ' ~ ', xFacet))),
               ifelse(!missing(yFacet),
                      g.plot <- g.plot + facet_grid(as.formula(paste(yFacet, ' ~ .'))),
                      ifelse(!missing(xFacet),
                             g.plot <- g.plot + facet_grid(as.formula(paste('. ~ ', xFacet))),
                             g.plot)))

        g.plot
}

facetHistDensity <- function (df, plotx, yFacet) {        
        df <- df %>%
                group_by_(yFacet) %>%
                mutate(Per = 1/length(unique(ResponseID))) %>%
                group_by_(plotx, yFacet) %>%
                summarize(Percent = sum(Per))
        
        g.plot <- ggplot(df, aes_string(x = plotx, y = 'Percent'))
        g.plot <- g.plot +
                geom_bar(fill = '#B34F89', stat = 'identity') +
                facet_grid(as.formula(paste(yFacet, '~ .'))) +
                theme(axis.title.x=element_blank(), 
                      axis.title.y=element_blank()) +
                scale_y_continuous(labels = percent) 
        g.plot
               
}

meltFactor <- function (df, factor, factorName, factorValue) {
        df.melt <- select(df, ResponseID, starts_with(factor), -ends_with('Other'), -ends_with('OtherComment')) %>%
                melt(id = 'ResponseID', variable.name = factorName, value.name = factorValue, na.rm=T)
        df.melt[ ,3] <- factor(df.melt[ ,3])
        df.melt
}

analysis/3_org.R
# subset all organizational characteristics

# melt/clean organization types
df.typesbase <- meltFactor(df, 'T.', 'OrgTypeCat', 'OrgType')
df.typesbase$OrgType <- revalue(df.typesbase$OrgType, c('Records Information Management' = 'Records Mgmt',
                                                        'IT/System administration' = 'IT'))
# melt/clean parent organization types
df.parentorgsbase <- meltFactor(df, 'PO.', 'ParentOrgCat', 'ParentOrg')
df.parentorgsbase$ParentOrg <- revalue(df.parentorgsbase$ParentOrg, c('Academic/Education' = 'Academic',
                                                                      'Corporate/For-profit' = 'Corporate',
                                                                      'County or Municipality' = 'Local',
                                                                      'Nonprofit Organization/Foundation' = 'Nonprofit'))

# melt/clean staff responsibilities
df.resp <- meltFactor(df, 'R.', 'RespCat', 'Resp')
df.resp$Resp <- revalue(df.resp$Resp, c('Paid full or part-time professional staff with practitioner experience' = 'Dedicated',
                                        'Digital preservation duties assigned to various staff as needed' = 'Split among Staff',
                                        'Digital preservation services provided by an external vendor' = 'Vendor',
                                        'Volunteers (full or part-time)' = 'Volunteer',
                                        'No staff person has digital preservation responsibilities' = 'No Staff'))
df.resp$Resp <- factor(df.resp$Resp, c('Dedicated', 'Split among Staff', 'Vendor', 'Volunteer', 'No Staff'))

# subset all single column org factors
df.org <- select(df, ResponseID, State, Timezone, StaffSize, AidAmount, Importance)

# melt/clean content formats
df.cformat <- meltFactor(df, 'CF.', 'ContentFormatCat', 'ContentFormat')
df.cformat$ContentFormat <- revalue(df.cformat$ContentFormat,
                                    c('Born-digital content created by and for your organization' = 'Born-digital',
                                      'Deposited digital materials managed for other individuals or institutions' = 'Deposited materials', 
                                      'Licensed digital content' = 'Licensed',
                                      'Reformatted material digitized from collections already held' = 'Digitized'))

# melt/clean content types
df.ctypes <- meltFactor(df, 'CT.', 'ContentTypeCat', 'ContentType')
df.ctypes$ContentType <- revalue(df.ctypes$ContentType, c('Architectural, CAD, and other design drawings' = 'CAD',
                                                          'Audiovisual files' = 'AV',
                                                          'Digital image files' = 'Images',
                                                          'Geographic information files' = 'GIS',
                                                          'PDF files' = 'PDF',
                                                          'Research data files' = 'Research data'))

# count responses by organzation type and parent organization
g.orgtype <- facetHist(merge(df.typesbase, df.parentorgsbase), 'OrgType', 'ParentOrg')

# filter organizations and parent organizations to largest members
df.parentorgs <- filter(df.parentorgsbase, ParentOrg %in% c('Academic', 'Nonprofit', 'State'))
df.types <- filter(df.typesbase, OrgType %in% c('Archive', 'Library', 'Museum'))

# compare staff size across organzation types
g.orgstaff <- facetHistDensity(merge(df.types, df.org), 'StaffSize', 'OrgType')

# compare staff size in libraries
g.libstaff <- facetHistDensity(merge(merge(df.types[df.types$OrgType == 'Library', ], df.org), df.parentorgsbase), 'StaffSize', 'ParentOrg')

# count minimum staff represented by libraries
df.aclib.staff <- merge(df.types[df.types$OrgType == 'Library', ],
                        df.parentorgs[df.parentorgs$ParentOrg == 'Academic', ]) %>%
        merge(df.org) %>%
        group_by(StaffSize) %>%
        summarize(count = length(ResponseID)) %>%
        select(count)

libstaff <- floor(sum(df.aclib.staff$count * c(1, 13, 25, 51, 201, 501))/sum(df.aclib.staff$count))

# compare digital preservation responsibility by staff size
g.respstaff <- facetHistDensity(merge(df.resp, df.org), 'Resp', 'StaffSize')

# compare digital preservation responsibility by organization type
g.resporg <- facetHistDensity(merge(df.resp, df.types), 'Resp', 'OrgType')

# compare content type by parent organization
g.ctypeparent <- facetHistDensity(merge(df.parentorgs, df.ctypes), 'ContentType', 'ParentOrg')

# compare content format by parent organization
g.cformparent <- facetHistDensity(merge(df.parentorgs, df.cformat), 'ContentFormat', 'ParentOrg')

# compare content type by organization type
g.ctypeorg <- facetHistDensity(merge(df.types, df.ctypes), 'ContentType', 'OrgType')

# compare content format by organization type
g.cformorg <- facetHistDensity(merge(df.types, df.cformat), 'ContentFormat', 'OrgType')

# compare content type by staff size
g.ctypestaff <- facetHistDensity(merge(df.org, df.ctypes), 'ContentType', 'StaffSize')

# compare content format by staff size
g.cformstaff <- facetHistDensity(merge(df.org, df.cformat), 'ContentFormat', 'StaffSize')

# count aid for training
g.aid <- ggplot(df.org, aes(x = AidAmount)) + geom_histogram(fill = '#B34F89')

# filter "don't know"s and least used responses
df.orgaid <- filter(df.org, AidAmount %in% c('$0-250', '$251-500', '$501-750', '$751-1000', '$1001-3000'))

# compare aid to digital preservation responsibility
g.aidresp <- facetHistDensity(merge(df.resp, df.orgaid), 'AidAmount', 'Resp')

# compare aid to staff size
g.aidstaff <- facetHistDensity(merge(df.types, df.orgaid), 'AidAmount', 'StaffSize')